src/hgraph/_wiring/_graph_builder.py

```diff
@@ -52,53 +53,78 @@ def create_graph_builder(sink_nodes: tuple["WiringNodeInstance"], supports_push_
+        edges.update(input_edges)
+        node_map[wiring_node] = ndx
 
     return GraphBuilderFactory.make(node_builders=tuple[NodeBuilder, ...](node_builders),
                                     edges=tuple[Edge, ...](
                                         sorted(edges,
                                                key=lambda e: (
                                                    e.src_node, e.dst_node, e.output_path, e.input_path))))
+
+
+def toposort(nodes: typing.Sequence["WiringNodeInstance"],
+             supports_push_nodes: bool = True) -> typing.Sequence["WiringNodeInstance"]:
+    mapping: dict["WiringNodeInstance", set["WiringNodeInstance"]] = defaultdict(set)
+    nodes_to_process: deque["WiringNodeInstance"] = deque(nodes)
+    source_nodes = set()
+    processed_nodes = dict["WiringNodeInstance", int]()
+    # Build node adjacency matrix and collect source nodes.
+    while len(nodes_to_process) > 0:
+        to_node = nodes_to_process.popleft()
+        if to_node in processed_nodes:  # This could be done better
+            continue  # This node has already been processed
+        else:
+            processed_nodes[to_node] = 1
+        ts_nodes = [n.node_instance for n in to_node.inputs.values() if isinstance(n, WiringPort)]
+        ts_nodes.extend(n for n in to_node.rank_marker.values())
+        for from_node in ts_nodes:
+            mapping[from_node].add(to_node)
+            if from_node.is_source_node:
+                source_nodes.add(from_node)
+                processed_nodes[from_node] = 1  # Since we are not going to add for processing
+            else:
+                nodes_to_process.append(from_node)
+    # Rank nodes
+    nodes_to_process.extend(source_nodes)
+    max_rank = 0
+    while len(nodes_to_process) > 0:
+        from_node = nodes_to_process.pop()
+        if not supports_push_nodes and from_node.resolved_signature.node_type is NodeTypeEnum.PUSH_SOURCE_NODE:
+            raise CustomMessageWiringError(
+                f'Node: {from_node.resolved_signature} is a push source node, '
+                f'but this graph does not support push nodes.')
+        if from_node.resolved_signature.node_type is WiringNodeType.PUSH_SOURCE_NODE:
+            # We re-set to 0 if this is a push source node to ensure they all left-align
+            processed_nodes[from_node] = 0
+        for to_node in mapping[from_node]:
+            processed_nodes[to_node] = max(processed_nodes[to_node], processed_nodes[from_node] + 1)  # increment
+            nodes_to_process.append(to_node)
+            max_rank = max(max_rank, processed_nodes[to_node])
+
+    # Set sink nodes to be max rank
+    for node in nodes:
+        processed_nodes[node] = max_rank
+
+    # Sort nodes by rank
+    result = [node for _, node in sorted((rank, node) for node, rank in processed_nodes.items()) if not node.is_stub]
+    if not all(n.rank <= n_1.rank for n, n_1 in zip(result[:-1], result[1:])):
```

Comment:
I have this to perform like-for-like comparisons, once the ranking logic is good will remove this and all the current rank properties from the code.